### ISSUE TITLE
No package file generated for a hosted repo = no repo

### ISSUE THREAD

**Issue by screig** [sonatype-nexus-community/nexus-repository-r#10]
A proxy repo generates a PACAKAGES.gz file, (but no PACKAGES’, or ‘PACKAGES.rds files)

![image](https://user-images.githubusercontent.com/4292074/29271207-4eb83382-80f2-11e7-96de-4e2f74809a24.png)

however a hosted repo does not have any packages files:

![image](https://user-images.githubusercontent.com/4292074/29271237-6c0d5cdc-80f2-11e7-94aa-5e62e6928183.png)

which means no R clients cannot access/use this repo. I don't think this is a problem with my package as  nexus can interrogate the package:

![image](https://user-images.githubusercontent.com/4292074/29271766-0e97c846-80f5-11e7-91ab-8789fa9f21cd.png)

It also seems that the other two package files (in the proxy repo)

> [the ‘PACKAGES’, ‘PACKAGES.gz’ and ‘PACKAGES.rds’ files](https://stat.ethz.ch/R-manual/R-devel/library/tools/html/writePACKAGES.html)

are maybe getting suppressed as they are not recognized by the code, (from the log file)....

```
2017-08-14 14:02:51,856+0200 INFO  [qtp2138984216-62] *UNKNOWN org.sonatype.nexus.repository.httpclient.internal.HttpClientFacetImpl - Repository status for r-test changed from READY to AVAILABLE - reason n/a for n/a
2017-08-14 14:02:51,966+0200 WARN  [qtp2138984216-62] *UNKNOWN org.sonatype.nexus.repository.httpbridge.internal.ViewServlet - Failure servicing: GET /repository/r-test/src/contrib/PACKAGES.rds
java.lang.IllegalStateException: Unexpected file extension for file: PACKAGES.rds
```

In the same vain I tried to create a 'group' repo adding in the hosted and proxy one. This did not work either, the package file was just copied over from the proxy repo, and so the file did not detail the additional packages located on the hosted. I also get the impression if I took two proxy repo and merged them in to a group repo that the package file would not represent the unique collection of all packages from both repos?

**Comment by screig** [comment]
I had a look through the code (i'm no java programmer!) I cant see anything to generate any of the of the PACKAGE files. Is this correct?

**Comment by DarthHater** [comment]
@screig , we don't generate a physical file for Hosted. If the route is hit, what we do is generate it on the fly (for now):

https://github.com/sonatype/nexus-repository-r/blob/master/src/main/java/org/sonatype/nexus/repository/r/internal/RHostedFacetImpl.java#L75

You can see the method that does all of that above.

We are definitely only generating files that match PACKAGES.gz as a name:

https://github.com/sonatype/nexus-repository-r/blob/master/src/main/java/org/sonatype/nexus/repository/r/internal/RRecipeSupport.groovy#L112

I was blissfuly unaware of the just PACKAGES or PACKAGES.rdl, but I'd love to know more about them. Can you link to a few examples?

**Comment by screig** [comment]
ok, bit confused by

> We are definitely only generating files that match PACKAGES.gz as a name:

An R repo unlike say a nuget or python repo works by the client interrogating the repo generated PACKAGE/PACKAGES.gz files (see [here](https://cran.r-project.org/doc/manuals/R-admin.html#Setting-up-a-package-repository)): 

_

> Each terminal directory must also contain a PACKAGES file. This can be a concatenation of the DESCRIPTION files of the packages separated by blank lines, but only a few of the fields are needed. The simplest way to set up such a file is to use function write_PACKAGES in the tools package, and its help explains which fields are needed. Optionally there can also be PACKAGES.rds and PACKAGES.gz files, downloaded in preference to PACKAGES. (These files will be smaller: PACKAGES.rds is used only from R 3.4.0. If you have a mis-configured server that does not report correctly non-existent files you may need these files.)

_

So without at least one of these files, correctly describing the repo, the R client sees nothing. In this trivial case if a user uploads a package to a hosted repo, at least a PACKAGE file needs to constructed and placed in the same sub-dir. The construction and deployment of the file needs to be done in isolation to other upload process.  

Note that the main CRAN repo will be calling [this function ](https://stat.ethz.ch/R-manual/R-devel/library/tools/html/writePACKAGES.html) for every sub directory on each insertion/deletion of a package, generating the three files and overwriting them in the repo.

at the top level of the repo we have 

- binary or source
- then os, windows etc
- then R version number

so eg

https://cran.r-project.org/bin/windows/contrib/

![image](https://user-images.githubusercontent.com/4292074/29284959-a60fac78-8124-11e7-8686-cf8acc4b03b4.png)

you can look through this sub folders and see in all cases these files exist and are different for each sub-dir.

![image](https://user-images.githubusercontent.com/4292074/29284993-c7f25430-8124-11e7-8860-0b515906d945.png)

**Note these file are all the client looks at!** So each time a package is added/removed the PACKAGES files need to be regenerated server side. R provides the above function to generate them, and I think your plugin should be reaching the same result. As you have already written the code to read the DESCRIPTION files you have done the hard work. However you do need to filer the selection of files to the given sub-directory you cant use everything. As well as that these function needs to be kicked off each time a package is added (a zip or tar file). But not kicked off when the PACKAGES, PACKAGES.gz or PACKAGES.rds file are added (circular process).

**Comment by DarthHater** [comment]
Right, what I was saying is we generate them on the fly, we don't store them in the repo. We did this to avoid cases where the PACKAGES.gz might get out of date as new packages are added. Right now, the path itself that generates them is just /blah/blah/blah/PACKAGES.gz, so anything that is PACKAGES or PACKAGES.rds is not generating one (and will require some additional work, duh :))

Thanks for the plethora of info, I'll see what we can do to take a look (and if anyone else in the community wants to, feel free to speak up)

**Comment by screig** [comment]
ps because of the different nature of how the client interacts with this repo, I think a group repo will be very difficult to achieve, you may want to disable that.  A nuget client essentially looks at the folder or folders containing the packages and can filter for those relevant to the framework, os, whatever. Adding a group of repos is easy as we simply tell the client to look in all these x places before filtering. With R the equivalent functionality is achieved by adding each repo on the client side, see [here](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/setRepositories.html).

**Comment by screig** [comment]
Are you certain that even the PACKAGES.gz file is generated for a hosted repo? I dont think it is? How could I check

**Comment by DarthHater** [comment]
@screig an easy-ish way would be to use Charles (or in your case if you are using Windows, Fiddler) to watch network traffic. You should see calls from R to Nexus Repository, and be able to see PACKAGES.gz somewhere in there. I've asked a couple of other devs to take a look at this issue internally, too, so they may have suggestions.

Group is always the funnest part of implementing any repo format (since we have to merge metadata), but we do our best to do so. That's part of why we chose to do the on the fly creation of PACKAGES.gz, it was easier to do the merge that way at the time.

**Comment by fjmilens3** [comment]
@screig (and @DarthHater) just to see that it's even being generated I would recommend the following to confirm the presence or absence of the (lazily-generated) file. Note that these instructions are for Mac in particular, though are translatable to any *nix-style operating system.

First download a known R package. I went with `agricolae` but any would be sufficient:

    $ curl -O https://cran.r-project.org/src/contrib/agricolae_1.2-6.tar.gz

Upload this to your Nexus instance once you've created an R hosted repository, changing this url to have the current username, password, and repository name:

    $ curl http://admin:admin123@localhost:8081/repository/r-hosted/src/contrib/agricolae_1.2-6.tar.gz --upload-file agricolae_1.2-6.tar.gz

Now that you've done that and the upload has completed (you should be able to confirm through the UI among other ways), try to download the PACKAGES.gz for that location:

    $ curl -O http://localhost:8081/repository/r-hosted/src/contrib/PACKAGES.gz

You should get the file. Once you have the file, run the following to get its contents:

    $ gunzip -c PACKAGES.gz

You should then see something similar to the following as the contents of the PACKAGES.gz:

    Package: agricolae
    Version: 1.2-6
    Depends: R (>= 2.10)
    Imports: klaR, MASS, nlme, cluster, spdep, AlgDesign, graphics
    Suggests: RANN
    License: GPL
    NeedsCompilation: no

**Comment by screig** [comment]
Hi @fjmilens3 

Thanks that tip worked....

curl -O http://**myIPaddres**:**myPort**/repository/**my_repo_name**/bin/windows/contrib/3.4/PACKAGES.gz

I can see the issue now, is that I am testing with windows binaries and not source packages as you did. R by default goes off and looks for src packages. As such added the following two lines to my Rprofile file (if this helps anybody, in the future)

```
options(install.packages.check.source = "no")
options(pkgType = "win.binary")
```

That said I would still like to be able to see (somehow) the PACKAGES.gz file through the web interface, even if it is not physically stored next to the packages. I would imagine that it would be better if proxy and hosted repo _looked_ the same. The missing PACKAGES.gz file in the hosted when viewed relative to a proxy repo is confusing. Likewise could the following line

`$ gunzip -c PACKAGES.gz`

not be inserted in to the JAVA code (in JAVAnese of course) to make both the PACKAGES and PACKAGES.gz available dynamically and thus more in line with the R spec? 

**Thanks for everyone's help with these issues!**
Sean

**Comment by fjmilens3** [comment]
> That said I would still like to be able to see (somehow) the PACKAGES.gz file through the web interface, even if it is not physically stored next to the packages.

That might be harder than you might think, since our view displays actual assets in the system, and not "synthesized" assets (for lack of a better term) that are just served up on request from database content.  

I can think of some ways to get around that, none of them particularly aesthetically appealing, and there's also likely the "better" approach of rebuilding this file when changes occur to a particular path in the repository structure (similar to other formats, notably our recent Yum release). 

Much of this was released as working but still in need of perfection, hence our leveraging of contributors in the open-source community to the greatest extent practicable.

> Likewise could the following line...not be inserted in to the JAVA code (in JAVAnese of course) to make both the PACKAGES and PACKAGES.gz available dynamically and thus more in line with the R spec?

Of course it could, at least as an interim step before the above. We're already building the content, so what you'd do is just refactor the code so that you build out the uncompressed content and send it back for the PACKAGES file, and for the PACKAGES.gz just compress it as part of building or streaming the response. 

That said, a better long-term approach would be to build these files on change, similar to maintaining metadata for other formats including Yum and Maven.

**Comment by screig** [comment]
I would agree with you on building the files on a change may be a better way to go. But saying that I know [el Zilcho](https://theinfosphere.org/New_Justice_Team) about the Nexus code base.

**Comment by aornatovskyy** [comment]
Fix will be included in the Nexus Repository Manager  3.20.0 with the new version of the R plugin.

**Comment by aornatovskyy** [comment]
Implemented NXRM 3.20.0-04 and R 1.1.1
